File: backend/app/core/discovery.py
"""
Website discovery + validation (spec 7).

Two paths:
  1. The CSV supplied a website  -> validate it and confirm it belongs to
     this business.
  2. The CSV supplied nothing    -> optionally try a small set of candidate
     domains, each of which must PASS identity matching before it is
     accepted. A guess that cannot be corroborated is discarded, and the
     lead is recorded as no_website rather than given someone else's site.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional

from .fetcher import Fetcher
from .page import parse_html
from .urls import (
    candidate_domains,
    is_non_website_host,
    normalize_url,
    registrable_domain,
    score_identity,
    tlds_for_region,
)

# website_status values (spec 7)
STATUS_VALID = "valid"
STATUS_REDIRECTED = "redirected"
STATUS_UNAVAILABLE = "unavailable"
STATUS_BLOCKED = "blocked"
STATUS_MISMATCH = "mismatch"
STATUS_NOT_FOUND = "not_found"
STATUS_NOT_A_WEBSITE = "not_a_website"   # social / directory / link-in-bio URL
STATUS_NO_WEBSITE = "no_website"         # verified: none supplied, none found


@dataclass
class DiscoveryResult:
    website_original: str = ""
    website_final: str = ""
    status: str = STATUS_NO_WEBSITE
    source: str = ""                 # csv | discovered | none
    identity_confidence: Optional[float] = None
    identity_verdict: str = ""
    identity_signals: List[Dict[str, Any]] = field(default_factory=list)
    notes: List[str] = field(default_factory=list)
    error_code: str = ""
    error_message: str = ""
    redirect_chain: List[str] = field(default_factory=list)
    http_status: Optional[int] = None
    response_ms: Optional[int] = None
    candidates_tried: List[str] = field(default_factory=list)
    social_profile_url: str = ""

    @property
    def has_website(self) -> bool:
        return self.status in (STATUS_VALID, STATUS_REDIRECTED) and bool(self.website_final)


async def discover_website(
    fetcher: Fetcher,
    *,
    business_name: str,
    website_raw: str = "",
    phone_digits: Optional[List[str]] = None,
    city: str = "",
    state: str = "",
    postal_code: str = "",
    address: str = "",
    category: str = "",
    region: Optional[str] = None,
    enable_guessing: bool = True,
    min_confidence: float = 0.55,
    max_candidates: int = 4,
) -> DiscoveryResult:
    out = DiscoveryResult(website_original=(website_raw or "").strip())

    # ---------------- path 1: CSV supplied a URL -------------------------
    if website_raw and str(website_raw).strip():
        norm = normalize_url(website_raw)
        if not norm:
            out.status = STATUS_NOT_FOUND
            out.source = "csv"
            out.notes.append("The website value in the CSV could not be parsed as a URL.")
        else:
            is_profile, kind = is_non_website_host(norm)
            if is_profile:
                out.status = STATUS_NOT_A_WEBSITE
                out.source = "csv"
                out.social_profile_url = norm
                out.notes.append(
                    f"The supplied URL is a {kind.replace('_', ' ')}, not the business's own website."
                )
                return out

            verified = await _fetch_and_verify(
                fetcher, norm, business_name=business_name, phone_digits=phone_digits,
                city=city, postal_code=postal_code, address=address, category=category,
            )
            out.http_status = verified["http_status"]
            out.response_ms = verified["response_ms"]
            out.redirect_chain = verified["redirect_chain"]
            out.source = "csv"

            if verified["error_code"]:
                out.error_code = verified["error_code"]
                out.error_message = verified["error_message"]
                out.status = (
                    STATUS_BLOCKED if verified["error_code"] == "blocked" else STATUS_UNAVAILABLE
                )
                out.notes.append(verified["error_message"])
                return out

            ident = verified["identity"]
            out.identity_confidence = ident["confidence"]
            out.identity_verdict = ident["verdict"]
            out.identity_signals = ident["signals"]
            out.website_final = verified["final_url"]

            # A URL the user gave us is trusted as *theirs*; identity scoring
            # only downgrades it when the page clearly belongs elsewhere.
            if ident["confidence"] < 0.2 and ident["verdict"] == "no_match":
                out.status = STATUS_MISMATCH
                out.notes.append(
                    "The page shows no sign of this business (name, phone or address). "
                    "Treated as a possible mismatch and excluded from outreach."
                )
                return out

            same = registrable_domain(norm) == registrable_domain(verified["final_url"])
            out.status = STATUS_VALID if same else STATUS_REDIRECTED
            if not same:
                out.notes.append(
                    f"Redirects to a different domain: {registrable_domain(verified['final_url'])}"
                )
                is_profile2, kind2 = is_non_website_host(verified["final_url"])
                if is_profile2:
                    out.status = STATUS_NOT_A_WEBSITE
                    out.social_profile_url = verified["final_url"]
                    out.notes.append(
                        f"The domain now redirects to a {kind2.replace('_', ' ')}."
                    )
            return out

    # ---------------- path 2: nothing supplied ---------------------------
    if not enable_guessing:
        out.status = STATUS_NO_WEBSITE
        out.source = "none"
        out.notes.append("No website in the CSV and automatic discovery is disabled.")
        return out

    tlds = tlds_for_region(region)
    candidates = candidate_domains(business_name, tlds)[:max_candidates]
    out.candidates_tried = candidates
    if not candidates:
        out.status = STATUS_NO_WEBSITE
        out.source = "none"
        out.notes.append("The business name did not yield any usable domain candidate.")
        return out

    best: Optional[Dict[str, Any]] = None
    for cand in candidates:
        verified = await _fetch_and_verify(
            fetcher, cand, business_name=business_name, phone_digits=phone_digits,
            city=city, postal_code=postal_code, address=address, category=category,
        )
        if verified["error_code"]:
            continue
        ident = verified["identity"]
        if best is None or ident["confidence"] > best["identity"]["confidence"]:
            best = verified
        if ident["confidence"] >= 0.85:
            break  # good enough, stop hitting more hosts

    if best is None:
        out.status = STATUS_NO_WEBSITE
        out.source = "none"
        out.notes.append(
            f"No reachable website found for {len(candidates)} candidate domain(s)."
        )
        return out

    ident = best["identity"]
    out.identity_confidence = ident["confidence"]
    out.identity_verdict = ident["verdict"]
    out.identity_signals = ident["signals"]
    out.http_status = best["http_status"]
    out.response_ms = best["response_ms"]
    out.redirect_chain = best["redirect_chain"]

    if ident["confidence"] >= min_confidence:
        out.website_final = best["final_url"]
        out.status = STATUS_VALID
        out.source = "discovered"
        out.notes.append(
            f"Discovered by domain guess and confirmed at {int(ident['confidence'] * 100)}% "
            f"identity confidence."
        )
    else:
        out.status = STATUS_NO_WEBSITE
        out.source = "none"
        out.notes.append(
            f"A candidate domain responded but only matched this business at "
            f"{int(ident['confidence'] * 100)}% confidence (threshold "
            f"{int(min_confidence * 100)}%). Not attached, to avoid using another "
            f"company's website."
        )
    return out


async def verify_direct_website(fetcher: Fetcher, url: str) -> DiscoveryResult:
    """
    Path for a URL the user typed directly into the app to have it audited,
    rather than a CSV row being matched against a business identity. There is
    no business name, phone or address to corroborate against, so identity
    scoring does not apply here - the URL itself, explicitly supplied, is the
    ground truth. This never guesses a domain and never falls back to
    candidate matching (spec 7's guessing path is CSV-only).
    """
    out = DiscoveryResult(website_original=(url or "").strip(), source="manual")
    norm = normalize_url(url)
    if not norm:
        out.status = STATUS_NOT_FOUND
        out.notes.append("That URL could not be parsed.")
        return out

    is_profile, kind = is_non_website_host(norm)
    if is_profile:
        out.status = STATUS_NOT_A_WEBSITE
        out.social_profile_url = norm
        out.notes.append(
            f"This is a {kind.replace('_', ' ')}, not a standalone website that can be audited."
        )
        return out

    res = await fetcher.fetch(norm)
    out.http_status = res.status
    out.response_ms = res.elapsed_ms
    out.redirect_chain = res.redirect_chain
    if not res.ok:
        out.error_code = res.error_code
        out.error_message = res.error_message
        out.status = STATUS_BLOCKED if res.error_code == "blocked" else STATUS_UNAVAILABLE
        out.notes.append(res.error_message)
        return out

    final = res.final_url or norm
    out.website_final = final
    out.identity_confidence = 1.0
    out.identity_verdict = "explicit_url"
    out.identity_signals = [{"signal": "explicitly_supplied_url", "weight": 1.0, "detail": norm}]

    same = registrable_domain(norm) == registrable_domain(final)
    out.status = STATUS_VALID if same else STATUS_REDIRECTED
    if not same:
        out.notes.append(f"Redirects to a different domain: {registrable_domain(final)}")
        is_profile2, kind2 = is_non_website_host(final)
        if is_profile2:
            out.status = STATUS_NOT_A_WEBSITE
            out.social_profile_url = final
            out.notes.append(f"The domain now redirects to a {kind2.replace('_', ' ')}.")
    return out


async def _fetch_and_verify(
    fetcher: Fetcher,
    url: str,
    *,
    business_name: str,
    phone_digits: Optional[List[str]],
    city: str,
    postal_code: str,
    address: str,
    category: str,
) -> Dict[str, Any]:
    res = await fetcher.fetch(url)
    out: Dict[str, Any] = {
        "url": url,
        "final_url": res.final_url or url,
        "http_status": res.status,
        "response_ms": res.elapsed_ms,
        "redirect_chain": res.redirect_chain,
        "error_code": res.error_code,
        "error_message": res.error_message,
        "identity": {"confidence": 0.0, "verdict": "no_match", "signals": []},
        "html": "",
    }
    if not res.ok:
        return out

    page = parse_html(res.text, url, final_url=res.final_url, status=res.status)
    out["html"] = res.text
    out["identity"] = score_identity(
        business_name=business_name,
        url=res.final_url or url,
        page_title=page.title,
        page_text=page.text[:20000],
        phone_digits=phone_digits,
        city=city,
        postal_code=postal_code,
        address=address,
        category=category,
    )
    return out


def parked_domain_signals(text: str, title: str) -> List[str]:
    """Detect placeholder / parked / under-construction pages."""
    hits: List[str] = []
    blob = f"{title} {text[:4000]}".lower()
    patterns = {
        "domain is for sale": "The domain appears to be listed for sale.",
        "buy this domain": "The domain appears to be listed for sale.",
        "under construction": "The page says it is under construction.",
        "coming soon": "The page says it is coming soon.",
        "site is temporarily unavailable": "The host reports the site as unavailable.",
        "default web page": "This is the hosting provider's default page.",
        "welcome to nginx": "This is an unconfigured nginx default page.",
        "apache2 ubuntu default page": "This is an unconfigured Apache default page.",
        "future home of something quite cool": "This is a hosting placeholder page.",
        "this domain is parked": "The domain is parked.",
        "godaddy.com": "" ,
    }
    for needle, note in patterns.items():
        if needle in blob and note:
            hits.append(note)
    if len(re.sub(r"\s+", " ", text).strip()) < 200 and not hits:
        hits.append("The page contains almost no content.")
    return hits

File: cloudflare/src/core/discovery.ts
// Website discovery + validation. Ported from backend/app/core/discovery.py.
import { Fetcher } from "./fetcher";
import { parseHtml } from "./page";
import { candidateDomains, isNonWebsiteHost, normalizeUrl, registrableDomain, scoreIdentity, tldsForRegion, IdentityScore } from "./urls";

export const STATUS_VALID = "valid";
export const STATUS_REDIRECTED = "redirected";
export const STATUS_UNAVAILABLE = "unavailable";
export const STATUS_BLOCKED = "blocked";
export const STATUS_MISMATCH = "mismatch";
export const STATUS_NOT_FOUND = "not_found";
export const STATUS_NOT_A_WEBSITE = "not_a_website";
export const STATUS_NO_WEBSITE = "no_website";

export interface DiscoveryResult {
  website_original: string;
  website_final: string;
  status: string;
  source: string;
  identity_confidence: number | null;
  identity_verdict: string;
  identity_signals: IdentityScore["signals"];
  notes: string[];
  error_code: string;
  error_message: string;
  redirect_chain: string[];
  http_status: number | null;
  response_ms: number | null;
  candidates_tried: string[];
  social_profile_url: string;
}

export function hasWebsite(d: DiscoveryResult): boolean {
  return (d.status === STATUS_VALID || d.status === STATUS_REDIRECTED) && Boolean(d.website_final);
}

function base(websiteOriginal: string, source = ""): DiscoveryResult {
  return {
    website_original: websiteOriginal, website_final: "", status: STATUS_NO_WEBSITE, source,
    identity_confidence: null, identity_verdict: "", identity_signals: [], notes: [],
    error_code: "", error_message: "", redirect_chain: [], http_status: null, response_ms: null,
    candidates_tried: [], social_profile_url: "",
  };
}

/** Direct-URL audit path: the user explicitly supplied this URL, so identity
 * scoring does not apply - it is the ground truth. Never guesses a domain. */
export async function verifyDirectWebsite(fetcher: Fetcher, url: string): Promise<DiscoveryResult> {
  const out = base((url || "").trim(), "manual");
  const norm = normalizeUrl(url);
  if (!norm) {
    out.status = STATUS_NOT_FOUND;
    out.notes.push("That URL could not be parsed.");
    return out;
  }

  const { isProfile, kind } = isNonWebsiteHost(norm);
  if (isProfile) {
    out.status = STATUS_NOT_A_WEBSITE;
    out.social_profile_url = norm;
    out.notes.push(`This is a ${kind.replace(/_/g, " ")}, not a standalone website that can be audited.`);
    return out;
  }

  const res = await fetcher.fetch(norm);
  out.http_status = res.status;
  out.response_ms = res.elapsed_ms;
  out.redirect_chain = res.redirect_chain;
  if (!res.ok) {
    out.error_code = res.error_code;
    out.error_message = res.error_message;
    out.status = res.error_code === "blocked" ? STATUS_BLOCKED : STATUS_UNAVAILABLE;
    out.notes.push(res.error_message);
    return out;
  }

  const final = res.final_url || norm;
  out.website_final = final;
  out.identity_confidence = 1.0;
  out.identity_verdict = "explicit_url";
  out.identity_signals = [{ signal: "explicitly_supplied_url", weight: 1.0, detail: norm }];

  const same = registrableDomain(norm) === registrableDomain(final);
  out.status = same ? STATUS_VALID : STATUS_REDIRECTED;
  if (!same) {
    out.notes.push(`Redirects to a different domain: ${registrableDomain(final)}`);
    const { isProfile: isProfile2, kind: kind2 } = isNonWebsiteHost(final);
    if (isProfile2) {
      out.status = STATUS_NOT_A_WEBSITE;
      out.social_profile_url = final;
      out.notes.push(`The domain now redirects to a ${kind2.replace(/_/g, " ")}.`);
    }
  }
  return out;
}

interface Verified {
  finalUrl: string;
  httpStatus: number | null;
  responseMs: number | null;
  redirectChain: string[];
  errorCode: string;
  errorMessage: string;
  identity: IdentityScore;
  html: string;
}

async function fetchAndVerify(
  fetcher: Fetcher,
  url: string,
  ctx: { businessName: string; phoneDigits?: string[]; city?: string; postalCode?: string; address?: string; category?: string },
): Promise<Verified> {
  const res = await fetcher.fetch(url);
  const out: Verified = {
    finalUrl: res.final_url || url, httpStatus: res.status, responseMs: res.elapsed_ms,
    redirectChain: res.redirect_chain, errorCode: res.error_code, errorMessage: res.error_message,
    identity: { confidence: 0, verdict: "no_match", signals: [] }, html: "",
  };
  if (!res.ok) return out;

  const page = parseHtml(res.text, url, { final_url: res.final_url, status: res.status });
  out.html = res.text;
  out.identity = scoreIdentity({
    businessName: ctx.businessName, url: res.final_url || url, pageTitle: page.title,
    pageText: page.text.slice(0, 20000), phoneDigits: ctx.phoneDigits, city: ctx.city,
    postalCode: ctx.postalCode, address: ctx.address, category: ctx.category,
  });
  return out;
}

/** CSV-import identity-matching path. Reserved for API completeness — the
 * live UI only creates jobs via verifyDirectWebsite (Quick Audit). */
export async function discoverWebsite(
  fetcher: Fetcher,
  ctx: {
    businessName: string;
    websiteRaw?: string;
    phoneDigits?: string[];
    city?: string;
    state?: string;
    postalCode?: string;
    address?: string;
    category?: string;
    region?: string;
    enableGuessing?: boolean;
    minConfidence?: number;
    maxCandidates?: number;
  },
): Promise<DiscoveryResult> {
  const out = base((ctx.websiteRaw || "").trim());

  if (ctx.websiteRaw && ctx.websiteRaw.trim()) {
    const norm = normalizeUrl(ctx.websiteRaw);
    if (!norm) {
      out.status = STATUS_NOT_FOUND;
      out.source = "csv";
      out.notes.push("The website value in the CSV could not be parsed as a URL.");
      return out;
    }
    const { isProfile, kind } = isNonWebsiteHost(norm);
    if (isProfile) {
      out.status = STATUS_NOT_A_WEBSITE;
      out.source = "csv";
      out.social_profile_url = norm;
      out.notes.push(`The supplied URL is a ${kind.replace(/_/g, " ")}, not the business's own website.`);
      return out;
    }

    const verified = await fetchAndVerify(fetcher, norm, ctx);
    out.http_status = verified.httpStatus;
    out.response_ms = verified.responseMs;
    out.redirect_chain = verified.redirectChain;
    out.source = "csv";

    if (verified.errorCode) {
      out.error_code = verified.errorCode;
      out.error_message = verified.errorMessage;
      out.status = verified.errorCode === "blocked" ? STATUS_BLOCKED : STATUS_UNAVAILABLE;
      out.notes.push(verified.errorMessage);
      return out;
    }

    out.identity_confidence = verified.identity.confidence;
    out.identity_verdict = verified.identity.verdict;
    out.identity_signals = verified.identity.signals;
    out.website_final = verified.finalUrl;

    if (verified.identity.confidence < 0.2 && verified.identity.verdict === "no_match") {
      out.status = STATUS_MISMATCH;
      out.notes.push(
        "The page shows no sign of this business (name, phone or address). Treated as a possible mismatch and excluded from outreach.",
      );
      return out;
    }

    const same = registrableDomain(norm) === registrableDomain(verified.finalUrl);
    out.status = same ? STATUS_VALID : STATUS_REDIRECTED;
    if (!same) {
      out.notes.push(`Redirects to a different domain: ${registrableDomain(verified.finalUrl)}`);
      const { isProfile: isProfile2, kind: kind2 } = isNonWebsiteHost(verified.finalUrl);
      if (isProfile2) {
        out.status = STATUS_NOT_A_WEBSITE;
        out.social_profile_url = verified.finalUrl;
        out.notes.push(`The domain now redirects to a ${kind2.replace(/_/g, " ")}.`);
      }
    }
    return out;
  }

  if (!(ctx.enableGuessing ?? true)) {
    out.status = STATUS_NO_WEBSITE;
    out.source = "none";
    out.notes.push("No website in the CSV and automatic discovery is disabled.");
    return out;
  }

  const tlds = tldsForRegion(ctx.region);
  const candidates = candidateDomains(ctx.businessName, tlds).slice(0, ctx.maxCandidates ?? 4);
  out.candidates_tried = candidates;
  if (!candidates.length) {
    out.status = STATUS_NO_WEBSITE;
    out.source = "none";
    out.notes.push("The business name did not yield any usable domain candidate.");
    return out;
  }

  let best: Verified | null = null;
  for (const cand of candidates) {
    const verified = await fetchAndVerify(fetcher, cand, ctx);
    if (verified.errorCode) continue;
    if (!best || verified.identity.confidence > best.identity.confidence) best = verified;
    if (verified.identity.confidence >= 0.85) break;
  }

  if (!best) {
    out.status = STATUS_NO_WEBSITE;
    out.source = "none";
    out.notes.push(`No reachable website found for ${candidates.length} candidate domain(s).`);
    return out;
  }

  out.identity_confidence = best.identity.confidence;
  out.identity_verdict = best.identity.verdict;
  out.identity_signals = best.identity.signals;
  out.http_status = best.httpStatus;
  out.response_ms = best.responseMs;
  out.redirect_chain = best.redirectChain;

  const minConfidence = ctx.minConfidence ?? 0.55;
  if (best.identity.confidence >= minConfidence) {
    out.website_final = best.finalUrl;
    out.status = STATUS_VALID;
    out.source = "discovered";
    out.notes.push(`Discovered by domain guess and confirmed at ${Math.round(best.identity.confidence * 100)}% identity confidence.`);
  } else {
    out.status = STATUS_NO_WEBSITE;
    out.source = "none";
    out.notes.push(
      `A candidate domain responded but only matched this business at ${Math.round(best.identity.confidence * 100)}% confidence ` +
        `(threshold ${Math.round(minConfidence * 100)}%). Not attached, to avoid using another company's website.`,
    );
  }
  return out;
}
